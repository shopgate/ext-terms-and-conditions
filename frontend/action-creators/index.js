import { UPDATE_CHECKOUT_IS_ORDERABLE } from '../constants';

/**
 * Updates isOrderable for checkout button
 * @param {Bool} isOrderable isOrderable
 * @returns {Object} The dispatched action object.
 */
export const updateCheckoutIsOrderable = isOrderable => ({
  type: UPDATE_CHECKOUT_IS_ORDERABLE,
  isOrderable,
});
