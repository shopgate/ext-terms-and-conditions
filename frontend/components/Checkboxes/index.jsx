import React, { Component } from 'react';
import PropTypes from 'prop-types';
import CheckboxWrapper from './components/CheckboxWrapper';
import connect from './connector';
import styles from './style';

/**
 * Render Checkboxes based on configuration
 */
class Checkboxes extends Component {
  static propTypes = {
    cartProductIds: PropTypes.arrayOf(PropTypes.string).isRequired,
    checkValues: PropTypes.arrayOf(PropTypes.shape()).isRequired,
    getProducts: PropTypes.func.isRequired,
    setCheckoutIsOrderable: PropTypes.func.isRequired,
    termsToDisplay: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  };
  /**
  * Constructs
  * @param {Object} props props.
  */
  constructor(props) {
    super(props);
    this.state = {
      checkValues: this.props.checkValues,
    };
  }
  /**
   * Ensure Checkout is disabled on mount
   */
  componentDidMount() {
    const {
      getProducts,
      cartProductIds,
    } = this.props;
    this.updateCheckoutAllowed();
    getProducts(cartProductIds);
  }

  /**
   * Update state with new props
   * @param {Object} prevProps prevProps
   */
  componentDidUpdate(prevProps) {
    if (this.props.checkValues !== prevProps.checkValues) {
      this.updateCheckedValues();
      this.props.setCheckoutIsOrderable(false);
    }
  }

  /**
   * checkValues state with new props
   */
  updateCheckedValues() {
    this.setState((state, props) => ({
      checkValues: props.checkValues,
    }));
  }

  /**
   * Handle click of checkbox
   * @param {number} index index
   * @param {number} checkedIndex checkedIndec
   */
  handleClick = (index, checkedIndex) => {
    const updateCheckValues = [...this.state.checkValues];
    updateCheckValues[index].checkedValues[checkedIndex].checked =
      !this.state.checkValues[index].checkedValues[checkedIndex].checked;
    this.setState({ checkValues: updateCheckValues });
    this.updateCheckoutAllowed();
  }

  /**
   * Dispatch setCheckoutIsOrderable
   */
  updateCheckoutAllowed = () => {
    let orderable = true;
    this.state.checkValues.forEach((element) => {
      const { checkedValues = [] } = element || {};
      const falseItem = checkedValues.find(value => value.checked === false);
      if (falseItem) {
        orderable = false;
      }
    });
    this.props.setCheckoutIsOrderable(orderable);
  }

  /**
   * @returns {JSX}
   */
  render() {
    const {
      termsToDisplay,
    } = this.props;
    const {
      checkValues,
    } = this.state;
    const checkboxes = termsToDisplay.map((product, index) => {
      if (typeof checkValues[index] === 'undefined') {
        return null;
      }
      return (
        <CheckboxWrapper
          checkedValues={checkValues[index].checkedValues}
          product={product}
          key={index.toString()}
          wrapperIndex={index}
          handleClick={this.handleClick}
        />);
    }).filter(element => element !== null);
    return (
      <div className={styles.wrapper}>
        <div className={styles.content}>
          { checkboxes }
        </div>
      </div>
    );
  }
}

export default connect(Checkboxes);
