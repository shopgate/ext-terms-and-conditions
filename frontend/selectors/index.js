import { createSelector } from 'reselect';
import { getCartProducts } from '@shopgate/pwa-common-commerce/cart/selectors';
import { getProducts } from '@shopgate/pwa-common-commerce/product/selectors/product';
import getConfig from '../helpers/getConfig';

const { checkboxValues } = getConfig();

/**
 * @param {Object} state state
 * @returns {Object}
 */
const getExtensionsState = state => state.extensions;

const getExtensionState = createSelector(
  getExtensionsState,
  extensionsState => extensionsState['@shopgate/terms-and-conditions/isOrderableByTerms']
);

export const getTermsStatus = createSelector(
  getExtensionState,
  extensionState => extensionState.isOrderable
);

export const getCartProductIds = createSelector(
  getCartProducts,
  (cartProducts) => {
    if (!cartProducts) {
      return null;
    }
    return cartProducts.map(cartProduct => cartProduct.product.id);
  }
);

export const getTermProductIds = createSelector(
  getCartProductIds,
  getProducts,
  (cartProductIds, products) => {
    const productsToCheck = [];
    if (!cartProductIds) {
      return null;
    }
    cartProductIds.forEach((cartProductId) => {
      if (products.hasOwnProperty(cartProductId)) {
        const testBaseProductId = products[cartProductId].productData.baseProductId;
        const idToPush = (testBaseProductId || cartProductId);
        productsToCheck.push(idToPush);
      }
    });
    return productsToCheck;
  }
);

export const getTermsToDisplay = createSelector(
  getTermProductIds,
  (productIds) => {
    if (!productIds) {
      return null;
    }
    return checkboxValues.filter((value =>
      productIds.includes(value.displayOn) || value.displayOn === 'all products'));
  }
);

export const getCheckValues = createSelector(
  getTermsToDisplay,
  (terms) => {
    const checkedValues = [];
    if (!terms) {
      return null;
    }
    terms.forEach((product, index) => {
      checkedValues.push({
        displayFor: `${product.displayOn}`,
        checkedValues: [],
      });
      product.productCheckboxValues.forEach(() => {
        checkedValues[index].checkedValues.push({ checked: false });
      });
    });
    return checkedValues;
  }
);

