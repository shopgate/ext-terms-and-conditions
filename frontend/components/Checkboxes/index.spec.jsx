import React from 'react';
import { mount } from 'enzyme';
import { createMockStore } from '@shopgate/pwa-common/store';
import { Provider } from 'react-redux';
import Checkboxes from './index';
import fetchProductsById from '@shopgate/pwa-common-commerce/product/actions/fetchProductsById';
import { getTermsToDisplay, getCheckValues, getCartProductIds } from '../../selectors';
import { updateCheckoutIsOrderable } from '../../action-creators';

jest.mock('./components/CheckboxWrapper', () => () => <div>CheckboxWrapper</div>);

jest.mock('../../selectors', () => ({
  getTermsToDisplay: jest.fn().mockReturnValue([{}]),
  getCheckValues: jest.fn().mockReturnValue([{}]),
  getCartProductIds: jest.fn().mockReturnValue(['1']),
}));
const mockedGetProductsById = jest.fn();
jest.mock('@shopgate/pwa-common-commerce/product/actions/fetchProductsById', () => (...args) => mockedGetProductsById(...args));
const store = createMockStore();

describe('Checkboxes', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it('should render', () => {
    const component = mount((
      <Provider store={store}>
        <Checkboxes
          cartProductIds={['1']}
          checkValues={[{}]}
          getProducts={() => {}}
          setCheckoutIsOrderable={() => {}}
          termsToDisplay={[{}]}
        />
      </Provider>
    ));
    expect(component.html()).toBe(null);
  });
});
