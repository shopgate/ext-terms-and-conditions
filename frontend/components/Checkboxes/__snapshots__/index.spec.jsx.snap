// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkboxes should render checkboxwrapper 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(Checkboxes)>
    <Checkboxes
      cartProductIds={
        Array [
          "1",
        ]
      }
      checkValues={
        Array [
          Object {},
        ]
      }
      getProducts={[Function]}
      setCheckoutIsOrderable={[Function]}
      termsToDisplay={
        Array [
          Object {},
        ]
      }
    >
      <div
        className="css-yrz0vs"
      >
        <div
          className="css-vmv4gl"
        >
          <Component
            handleClick={[Function]}
            key="0"
            product={Object {}}
            wrapperIndex={0}
          >
            <div>
              CheckboxWrapper
            </div>
          </Component>
        </div>
      </div>
    </Checkboxes>
  </Connect(Checkboxes)>
</Provider>
`;
